helm-app-operator/pkg/helm/installer/installer.go

```diff
@@ -0,0 +1,348 @@
+	"k8s.io/helm/pkg/tiller/environment"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+)
+
+// Installer can install and uninstall Helm releases given a custom resource
+// which provides runtime values for the Chart.
+type Installer interface {
+	InstallRelease(u *unstructured.Unstructured) (*unstructured.Unstructured, error)
+	UninstallRelease(u *unstructured.Unstructured) (*unstructured.Unstructured, error)
+}
+
+// assert interface
+var _ Installer = installer{}
+
+const (
+
+	// HelmChartWatchesEnvVar is the environment variable for a YAML
+	// configuration file containing mappings of GVKs to helm charts.
+	// Use of this environment variable overrides the watch configuration
+	// provided by API_VERSION, KIND, and HELM_CHART, and allows users to
+	// configurable multiple watches, each with a different chart.
+	HelmChartWatchesEnvVar = "HELM_CHART_WATCHES"
+
+	// APIVersionEnvVar is the environment variable for the group and version
+	// to be watched using the format `<group>/<version>`
+	// (e.g. "example.com/v1alpha1").
+	APIVersionEnvVar = "API_VERSION"
+
+	// KindEnvVar is the environment variable for the kind to be watched. The
+	// value is typically singular and should be CamelCased (e.g. "MyApp").
+	KindEnvVar = "KIND"
+
+	// HelmChartEnvVar is the environment variable for the directory location
+	// of the helm chart to be installed for CRs that match the values for the
+	// API_VERSION and KIND environment variables.
+	HelmChartEnvVar = "HELM_CHART"
+
+	// OperatorNameEnvVar is the environment variable for the operator name,
+	// which is used in the release name for helm releases. If not set, a
+	// default value will be used.
+	OperatorNameEnvVar = "OPERATOR_NAME"
+
+	defaultOperatorName         = "helm-app-operator"
+	defaultHelmChartWatchesFile = "/opt/helm/watches.yaml"
+)
+
+var (
+	operatorName = defaultOperatorName
+)
+
+func init() {
+	setOperatorName()
+}
+
+// installer is an implementation of the Installer interface, which is used to
+// reconcile CR updates for GVKs registered as helm apps.
+type installer struct {
+	tillerKubeClient *kube.Client
+	storageBackend   *storage.Storage
+	chartDir         string
+}
+
+// watch holds data used to create a mapping of GVK to helm chart.
+// The mapping is used to compose a helm app operator.
+type watch struct {
+	Version string `yaml:"version"`
+	Group   string `yaml:"group"`
+	Kind    string `yaml:"kind"`
+	Chart   string `yaml:"chart"`
+}
+
+// NewFromEnv returns a map of installers based on configuration provided in
+// the environment.
+func NewFromEnv(tillerKubeClient *kube.Client, storageBackend *storage.Storage) (map[schema.GroupVersionKind]Installer, error) {
+	// Order of precedence:
+	// 1. HELM_CHART_WATCHES
+	// 2. Default watches file
+	// 3. API_VERSION, KIND, and HELM_CHART
+
+	m := map[schema.GroupVersionKind]Installer{}
+
+	// Determine the watches file to use
+	watchesFile, ok := os.LookupEnv(HelmChartWatchesEnvVar)
```

Comment:
The HelmChartWatchesEnvVar  could be set to empty, `HELM_CHART_WATCHES =""`.
In which case `ok=true` and `watchesFile=""`. We'll end up trying to load the watches file in that case right?

Reply:
We would, in which case there would be an error returned by NewFromWatches. I figured that if the environment variable is set (regardless of its value), we should use it since the user explicitly set it.

I could go either way on this though. We could also fall back to the default file for the empty string case.

Reply:
Added a commit make the logic in this section of code more obvious.